maint(fix): Add deprecation warning for accidental imports

oscarbenjamin opened the pull request:
In https://github.com/sympy/sympy/pull/18245 a number of modules were
removed from the list of names imported by

   from sympy import *

This adds those names back although with most of them wrapped in an
object that emits DeprecationWarning each time it is accessed.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs

See the discussion on the mailing list:
https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/sympy/HYo616T3EwQ/g-z6ZrNKAgAJ

#### Brief description of what is fixed or changed

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
maint(fix): Add deprecation warning for accidental imports

In https://github.com/sympy/sympy/pull/18245 a number of modules were
removed from the list of names imported by

   from sympy import *

This adds those names back although with most of them wrapped in an
object that emits DeprecationWarning each time it is accessed.

oscarbenjamin commented:
Does anyone want to review this?

CC @moorepants @asmeurer

asmeurer commented:
Is the idea when the deprecation ends to continue to also have top-level modules be removed as well? I think they should be.

asmeurer commented on sympy/__init__.py:
```diff
@@ -496,3 +496,171 @@ def __sympy_debug():
+#=======================================================================
+
+
+class DeprecatedImportModule:
```
This should subclass types.ModuleType.

oscarbenjamin replied:
What difference does subclassing ModuleType make?

asmeurer replied:
It makes certain things that operate on modules work correctly. For example, `help()` gives special output on modules (actually that is wrong anyway, because this doesn't inherit the docstring properly). 

I also wonder if this could be an issue

```py
>>> from sympy import ode as ode1
>>> from sympy.solvers import ode as ode2
>>> ode1 == ode2
False
```

It might or might not be an issue, I'm not sure. I guess if this turns out to cause problems, we can do a patch release that removes it.

oscarbenjamin replied:
I think I'd prefer the help to be a message saying that you haven't imported the module in the right way (if there is any need for help at all).

The purpose of providing these deprecated module imports is just to keep code that possibly depends on importing this way chugging along but with warnings. I don't see any need to provide any other functionality.

moorepants commented on sympy/__init__.py:
```diff
@@ -496,3 +496,171 @@ def __sympy_debug():
+    'multipledispatch',
+    'ntheory',
+    'parsing',
+    # 'physics',
```
Maybe remove if this wasn't imported  before?

moorepants commented on sympy/__init__.py:
```diff
@@ -496,3 +496,171 @@ def __sympy_debug():
+# __init__.py file. The plan is to remove them but for now they remain
+# importable but will give a deprecation warning when used.
+#
+#=======================================================================
```
I don't understand the purpose of have this comment block repeated. It refers to the names below, but there is nothing below?

oscarbenjamin replied:
There's a bug list of names in `_DEPRECATED_IMPORTS` below. The difference between these and those that are listed above is that these are deprecated and will give a deprecation warning. The ones above (in this PR currently) are not deprecated and will give no warning when used.

moorepants commented:
As I understand this, some subset of all sympy modules were imported with `from sympy import *`. Those modules were then available in the namespace and functions, classes, and variables could be accessed with `module.function()` for example. So, with this warning you can still import and use the modules but you get a warning when doing so. I don't quite understand how you've implemented it, but the result seems very good.

It also seems you could do submodule star imports:

```
from sympy.core import *
add.Add()
```

I assume those will not get deprecation warnings with this?

asmeurer commented:
I think they don't but the mechanism could be added. However, my opinion is that submodule star imports should really be supported except for submodules that aren't already included in the main `sympy/__init__.py`.

oscarbenjamin commented:
> Is the idea when the deprecation ends to continue to also have top-level modules be removed as well? I think they should be.

I don't see any way to do that with a deprecation period so if we're removing them them from star import we might as well just do it now.

oscarbenjamin commented:
> I assume those will not get deprecation warnings with this?

This only gives deprecation warnings for modules that were previously importable from top level.

For example in sympy 1.5 you can do:
```
>>> from sympy.combinatorics import *
>>> prufer
<module 'sympy.combinatorics.prufer' from '/Users/enojb/current/sympy/sympy/sympy/combinatorics/prufer.py'>
```
With 1.6 the name prufer would not be imported. This PR does not change that.

asmeurer commented:
> I don't see any way to do that with a deprecation period so if we're removing them them from star import we might as well just do it now.

If we are going to do a deprecation I think it is better to remove them together, even if we can't issue a warning for some of the cases.

oscarbenjamin committed:
maint(fix): Add deprecation warning for accidental imports

This commit clarifies the comments about what will or will not be
importable in future.

oscarbenjamin committed:
maint(fix): Add deprecation warning for accidental imports

Clarify comments in __init__.py add docstring for DeprecatedImportModule

oscarbenjamin commented:
@moorepants @asmeurer does this look good?

If so I'll merge it and along with a fix for #19326 I can put out a second release candidate.

moorepants commented:
I'm good with this, as long as it gives a warning on the most likely cases. Thank you for addressing my concern.
